view/src/main/java/org/xcolab/view/pages/profile/beans/BadgeBean.java

```diff
@@ -1,82 +1,69 @@
-    private final long userId;
-    private final List<Badge> badges;
+
+    private final List<Badge> badges = new ArrayList<>();
 
     public BadgeBean(long userId) {
-        this.userId = userId;
-        badges = new ArrayList<>();
-        fetchBadges();
+        populateBadges(userId);
     }
 
-    private void fetchBadges() {
-        for(Proposal p : ProposalClientUtil.getMemberProposals(userId)) {
-            try {
-                final ContestPhaseRibbonType ribbonType = getRibbonType(p);
+    private void populateBadges(long userId) {
+        for (Proposal proposal : ProposalClientUtil.getMemberProposals(userId)) {
+            getRibbonAttribute(proposal)
+                    .ifPresent(ribbonAttribute -> {
+                        final ContestPhaseRibbonType ribbonType = ContestClientUtil
+                                .getContestPhaseRibbonType(ribbonAttribute.getNumericValue());
 
-                if (ribbonType != null  && ribbonType.getRibbon() > 0) {
-                    Contest contest = ProposalClientUtil.getCurrentContestForProposal(p.getProposalId());
-                    String proposalTitle = ProposalAttributeClientUtil
-                            .getProposalAttribute(p.getProposalId(), ProposalAttributeKeys.NAME, 0L)
-                            .getStringValue();
-                    badges.add(new Badge(ribbonType, p, proposalTitle, contest));
-                }
-            } catch (ContestNotFoundException e) {
-                _log.warn("Could not add badge to user profile view for userId: {} and proposalId: {}",
-                        p.getProposalId(), e);
-            }
+                        if (ribbonType.getRibbon() > 0) {
+                            final ContestPhase phase = ContestClientUtil.getContestPhase(
+                                    ribbonAttribute.getContestPhaseId());
+                            badges.add(new Badge(ribbonType, proposal, proposal.getName(),
+                                    phase.getContest()));
+                        }
+                    });
         }
     }
 
-    private ContestPhaseRibbonType getRibbonType(Proposal p) {
-        ContestPhaseRibbonType contestPhaseRibbonType = null;
-        List<Long> phasesForProposal = ProposalPhaseClientUtil.getContestPhasesForProposal(p.getProposalId());
-
-        Date phaseStartDate = p.getCreateDate();
-        for (Long phaseId : phasesForProposal) {
-            final ProposalContestPhaseAttribute ribbonAttribute =
-                    ProposalPhaseClientUtil.getProposalContestPhaseAttribute(p.getProposalId(),
-                            phaseId, ProposalContestPhaseAttributeKeys.RIBBON);
-            if (ribbonAttribute != null) {
-                long typeId = ribbonAttribute.getNumericValue();
-
-                ContestPhase contestPhase = ContestClientUtil.getContestPhase(phaseId);
+    private Optional<ProposalContestPhaseAttribute> getRibbonAttribute(Proposal proposal) {
```

Comment:
Very clever and elegant use of the functional style programming to significantly reduce the lines of code for this method. I like it a lot!

I would make a couple of changes to make the method easier to read:
* rename it to `getLatestRibbonAttribute` or something to allow the reader to understand what this method does exactly.
* change the `map` and `max` functions to inline lambdas and define static methods that take one parameter:

``` 
return phasesForProposal.stream()
    .map(phase -> getRibbonAttribute(phase))
    .filter(Objects::nonNull)
    .max(Comparator.comparing(attribute -> getStartDate(attribute)));
```

Reply:
Thanks! I like your ideas, that's a nice way to improve the readability by allowing the parameters to be named 👍 And you're right about the method name, it wasn't explaining why we compared dates yet. All done!